packages/queued-request-controller/src/QueuedRequestController.test.ts

```diff
@@ -33,176 +34,654 @@ describe('QueuedRequestController', () => {
+        const firstRequest = controller.enqueueRequest(
+          { ...buildRequest(), origin: 'https://firstorigin.metamask.io' },
+          () => new Promise((resolve) => setTimeout(resolve, 10)),
+        );
+        // ensure first request skips queue
+        expect(controller.state.queuedRequestCount).toBe(0);
+        const secondRequestNext = jest
+          .fn()
+          .mockImplementation(
+            () => new Promise((resolve) => setTimeout(resolve, 100)),
+          );
+        const secondRequest = controller.enqueueRequest(
+          { ...buildRequest(), origin: 'https://secondorigin.metamask.io' },
+          secondRequestNext,
+        );
+        // ensure test starts with one request queued up
+        expect(controller.state.queuedRequestCount).toBe(1);
+        expect(secondRequestNext).not.toHaveBeenCalled();
+
+        await firstRequest;
+        await expect(secondRequest).rejects.toThrow(switchError);
+      });
+
+      it('correctly processes the next item in the queue', async () => {
+        const switchError = new Error('switch error');
+        const { messenger } = buildControllerMessenger({
+          networkControllerGetState: jest.fn().mockReturnValue({
+            ...cloneDeep(defaultNetworkState),
+            selectedNetworkClientId: 'selectedNetworkClientId',
+          }),
+          networkControllerSetActiveNetwork: jest
+            .fn()
+            .mockRejectedValue(switchError),
+          selectedNetworkControllerGetNetworkClientIdForDomain: jest
+            .fn()
+            .mockImplementation((origin) =>
+              origin === 'https://secondorigin.metamask.io'
+                ? 'differentNetworkClientId'
+                : 'selectedNetworkClientId',
+            ),
+        });
+        const options: QueuedRequestControllerOptions = {
+          messenger: buildQueuedRequestControllerMessenger(messenger),
+        };
+        const controller = new QueuedRequestController(options);
+        const firstRequest = controller.enqueueRequest(
+          { ...buildRequest(), origin: 'https://firstorigin.metamask.io' },
+          () => new Promise((resolve) => setTimeout(resolve, 10)),
+        );
+        // ensure first request skips queue
+        expect(controller.state.queuedRequestCount).toBe(0);
+        const secondRequestNext = jest
+          .fn()
+          .mockImplementation(
+            () => new Promise((resolve) => setTimeout(resolve, 100)),
+          );
+        const secondRequest = controller.enqueueRequest(
+          { ...buildRequest(), origin: 'https://secondorigin.metamask.io' },
+          secondRequestNext,
+        );
+        const thirdRequestNext = jest
+          .fn()
+          .mockImplementation(
+            () => new Promise((resolve) => setTimeout(resolve, 100)),
+          );
+        const thirdRequest = controller.enqueueRequest(
+          { ...buildRequest(), origin: 'https://thirdorigin.metamask.io' },
```

Comment:
can this be firstorigin instead? feels weird to introduce a third origin here.
```suggestion
          { ...buildRequest(), origin: 'https://firstorigin.metamask.io' },
```

Reply:
Hmm, I guess it could be. The reasoning was that the first request just served to ensure the second was queued. The second is the important part of the test setup: that the test starts with 1 queued batch.

The third request just had to be anything other than the second origin. I used a third origin here so that it was clear that they didn't need to match.